### PR TITLE
[SLM] Add support for Baichuan2 architecture

### PR DESCRIPTION
**Pull request by tlopex**
Baichuan Intelligence Inc. has released a new big open-source language model called Baichuan 2. It was trained on a dataset with 2.6 trillion tokens and has performed the best in tests for both Chinese and English compared to other models of its size. Because it works so well, adding Baichuan 2's architecture to MLC-LLM could be a smart move to practice.

### CODE REVIEW

**Comment by tlopex** [comment]
cc @tqchen

File: python/mlc_chat/model/baichuan/baichuan_model.py [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        v = self.v_cache.view(t)
+        # Step 4. Compute softmax(Q @ K^T / sqrt(d)) @ V
+        output = op_ext.attention(q, k, v, casual_mask=attention_mask)
+        # Step 5. Apply output projection
+        return self.o_proj(output)
+
+
+class BaichuanMLP(nn.Module):
+    def __init__(self, config: BaichuanConfig):
+        self.gate_proj = nn.Linear(config.hidden_size, config.intermediate_size, bias=False)
```
Comment (rickzx) [inline python/mlc_chat/model/baichuan/baichuan_model.py:119]: Here, we can combine the gate_proj and up_proj matmul similar to what we did in combing QKV in attention layer.
```
self.gate_up_proj = nn.Linear(
    in_features=config.hidden_size,
    out_features=2 * config.intermediate_size,
    bias=False,
)
self.down_proj = nn.Linear(self.intermediate_size, config.hidden_size, bias=False)
```

And in `forward`,
```
concat_x1_x2 = self.gate_up_proj(x)
x1, x2 = op.split(concat_x1_x2, 2, axis=-1)
return self.down_proj(self.act_fn(x1) * x2)
```

Also need to transform the parameters in the loader, similar to this one: https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_chat/model/llama/llama_loader.py#L65

**Review by rickzx — CHANGES_REQUESTED** [review]
Thanks for the contribution! Overall looks good. Please also address the lint errors in the CI

File: python/mlc_chat/model/baichuan/baichuan_model.py (hunk 1 above)
Comment (tlopex) [reply python/mlc_chat/model/baichuan/baichuan_model.py:119]: Got it. So it is usually better to fuse ops than use them independently?

Comment (tqchen) [reply python/mlc_chat/model/baichuan/baichuan_model.py]: yes,

**Comment by tqchen** [comment]
would be great to do some quick compile and confirm the model works as intended

**Comment by tlopex** [comment]
Sure! @tqchen 
Commands:
```
python -m mlc_chat convert_weight ./dist/models/Baichuan2-7B-Chat/ --quantization q4f16_1 -o dist/Baichuan2-7B-Chat-q4f16_1-MLC

# 1. gen_config: generate mlc-chat-config.json and process tokenizers
python -m mlc_chat gen_config ./dist/models/Baichuan2-7B-Chat/ --quantization q4f16_1 --conv-template chatml -o dist/Baichuan2-7B-Chat-q4f16_1-MLC/

# 2. compile: compile model library with specification in mlc-chat-config.json
python -m mlc_chat compile ./dist/Baichuan2-7B-Chat-q4f16_1-MLC/mlc-chat-config.json --device cuda -o dist/libs/Baichuan2-7B-Chat-q4f16_1-cuda.so
```
The screen shot of model running:
图片

**Comment by tqchen** [comment]
You can run chat through CLI https://llm.mlc.ai/docs/deploy/cli.html

**Comment by tqchen** [comment]
Please fix the CI errors, and address comments

**Comment by tqchen** [comment]
Please also read https://github.com/mlc-ai/mlc-llm/issues/1749, which is a followup item that we are performing right now to refactor models into the new PagedKVCache format.

Does not have to tie with this PR, but we should followup soon

**Comment by tlopex** [comment]
@tqchen Thanks!
Here is a screen shot of the compilation result after CLI
图片

I will address the errors and followup the item you mentioned.

File: python/mlc_chat/model/baichuan/baichuan_model.py (hunk 1 above)
Comment (tlopex) [reply python/mlc_chat/model/baichuan/baichuan_model.py]: I see. Thanks! I have already combined them in the code.

**Comment by tlopex** [comment]
I can't quite understand the CI error here:
```
Traceback (most recent call last):
  File "tests/python/integration/test_model_compile.py", line 12, in <module>
    from mlc_chat.model import MODEL_PRESETS
  File "/opt/conda/envs/ci-unittest/lib/python3.8/site-packages/mlc_chat/model/__init__.py", line 2, in <module>
    from .model import MODELS, Model
  File "/opt/conda/envs/ci-unittest/lib/python3.8/site-packages/mlc_chat/model/model.py", line 11, in <module>
    from .baichuan import baichuan_loader, baichuan_model, baichuan_quantization
ModuleNotFoundError: No module named 'mlc_chat.model.baichuan'
```
I wonder how can I fix it. The error is the same as my another PR.

**Comment by tqchen** [comment]
likely you need `__init__.py` under the baichuan folder

**Comment by tqchen** [comment]
great, will leave @rickzx to manage the pr

**Comment by rickzx** [comment]
Awesome, looks good to me

**Comment by gesanqiu** [comment]
Hi @tlopex , thanks for your amazing work. But I have two questions:
1. Why you use chatml conv_template, as far as I know, baichuan2 use `<reserved_106>{Question1}<reserved_107>{answer}<reserved_106>{Question2}<reserved_107>`. I found that if I generate with chatml conv_template, inference can stop properly, but can't stop generating if I use `<reserved_106>...<reserved_107>`.
2. I'm new to mlc-llm and tvm, have you considered supporting baichuan2-13b-chat, which relies on ALibi postion embedding. For a quick verfication, how about modifying the `_attention_mask()` to construct an `alibi_mask` and close flashinfer, so attention will use `_fallback()` then we can add the `alibi_mask` in the `_fallback()`?

**Comment by tlopex** [comment]
@gesanqiu Thanks for your questions!
For the first question, sorry I ignored that. I'll fix that in a few days.
For the second question, the community is continuously making progress on integrating and supporting new technologies. Your suggestion regarding the modification of the` _attention_mask()` to construct an `alibi_mask` and adjustments to the attention mechanism to utilize` _fallback()` with the inclusion of the `alibi_mask` are valuable insights. Please stay tuned for further updates.

**Comment by tqchen** [comment]
@tlopex please followup on https://github.com/mlc-ai/mlc-llm/issues/1749, we should migrate all models to new def

**Comment by tlopex** [comment]
@tqchen Got it. I'll update this model as well as the new one soon.

**Pull request merged**